ARCs/arc-0060.md

```diff
@@ -0,0 +1,449 @@
+
+This API was not designed to sign Algorand transactions or a group of transactions (see ARC-1 instead).
+
+## Backwards Compatibility
+
+This ARC is not backward compatible with [ARC-47](./arc-0047.md) Logic Signature Template.
+
+ARC-47 introduces the method `algo_templatedLsig` that takes a `LogicSignatureDescription` object and returns the signature of a logic signature program for delegation.
+
+An alternative using the ARC-60 `signData(signingData, metadata)` is given below, where:
+
+- `signingData.data` **MUST** specify:
+  - the `LogicSignatureDescription` object as detailed in ARC-47
+  - the `values` of the templated variables and the `hash` of the expected program
+  - the `ARC60Domain` must be set to the domain separator `"Program"`, as specified in the <a href="https://github.com/algorandfoundation/specs/blob/master/dev/crypto.md#domain-separation">Algorand specs</a>.
+
+- `metadata` **MUST** specify:
+  - an extended ARC-60 compatible JSON schema
+  - `ScopeType` is set to `LSIG`.
+
+The extended schema can be found in the `../assets/arc-0060` folder with the file [lsig-schema.json](../assets/arc-0060/lsig-schema.json).
+
+### Semantics for signing a Logic Signature
+
+- if the `metadata.scope` is set to `LSIG`:
+  - the `ARC60Domain` value **MUST** be equal to the byte array corresponding to the prefix `"Program"`.
+  - the JSON schema **MUST** define the `signingData.data.bytes` field as an array of objects, such as the `LogicSignatureDescription`, `values`, and `hash`.
+  - the wallet **MUST** follow the ARC-47 specification: compile the program from the received template and check its integrity with the `hash`.
+  - the compiled program **MUST** be used with the prefix to compute the signing message `msg`, as `msg="Program" || <compiled_lsig>`.
+
+## Test Cases
+
+N/A
+
+## Reference Implementation
+
+A reference implementation is available in the `../assets/arc-0060` folder.
+
+- [auth-schema.json](../assets/arc-0060/auth-schema.json) provides the ARC-60 JSON schema for `ScopeType=AUTH`.
+- [lsig-schema.json](../assets/arc-0060/lsig-schema.json) provides the ARC-60 JSON schema for `ScopeType=LSIG`.
+- [wallet.ts](../assets/arc-0060/wallet.ts) contains a simple TypeScript script showing how wallets can implement the `SignDataFunction` interface.
+
+## Security Considerations
+
+Users signing bytes can be tricked into signing malicious messages being exposed to man-in-the-middle attacks.
+
+Users must be aware of what they are signing and for what purpose. The wallet **MUST** always compute a new message dynamically according to the received `signingData` and `metadata`. This prevents the injection of malicious or misleading messages and provides a clear scope of the signing action.
+
+Wallets only support known `ScopeTypes` and schemas. It ensures that signing actions have a clear scope and mitigates the risk of producing unexpected signatures.
```

Comment:
I agree that wallets need to hard-code the `ScopeTypes` and schemas that they support. But this raises the question for me if we even need to include the `schema` in the `metadata` field?

Given the `ScopeType`, the wallet could look up its hard-coded schema and match the `signingData` against this schema.

Perhaps it's nice to have the schema also in the metadata to make the request self-descriptive. But in an implementation I'd likely ignore `metadata.schema` and match `signingData.bytes` against my hard-coded schema. Would that be a problem?